bokeh/util/structure.py

```diff
@@ -0,0 +1,315 @@
+"""Functions to create the graph of submodels of a model in networkx format,
+   and to draw that graph using bokeh so that one can explore the attributes of
+   submodels by clicking on the nodes.  Uses simple custom javascript callbacks so no server is necessary;
+   works in a jupyter notebook.
+
+   Suppose M is a bokeh model (i.e. a plot or figure).
+
+      - K = BokehStructureGraph(M) creates the graph, which is itself a bokeh model.
+      - K.show() calls show on the model to display it.
+      - K.model returns the model consisting of the structure graph and associated datatable.
+      - K.graph returns the directed graph in networkx format
+      - K.property_df returns the dataframe of models, submodels and their properties.
+"""
+
+# Standard library imports
+from itertools import permutations
+
+# External imports
+import networkx as nx
+import pandas as pd
```

Comment:
These are not hard dependencies of the project. It would be better for them to be protected imports inside any functions that use them, rather than at the top level.

Reply:
I've done this but I'm not sure I agree since although only certain functions use them, they are pretty fundamental to the way I've put this together so to do anything useful they are going to be imported.